Objects/longobject.c

```diff
@@ -4704,28 +4732,37 @@ long_bitwise(PyLongObject *a,
 static PyObject *
 long_and(PyObject *a, PyObject *b)
 {
-    PyObject *c;
     CHECK_BINOP(a, b);
-    c = long_bitwise((PyLongObject*)a, '&', (PyLongObject*)b);
-    return c;
+    PyLongObject *x = (PyLongObject*)a;
+    PyLongObject *y = (PyLongObject*)b;
+    if (IS_MEDIUM_VALUE(x) && IS_MEDIUM_VALUE(y)) {
+        return _PyLong_FromSDigit(medium_value(x) & medium_value(y));
+    }
+    return long_bitwise(x, '&', y);
 }
 
 static PyObject *
 long_xor(PyObject *a, PyObject *b)
 {
-    PyObject *c;
     CHECK_BINOP(a, b);
-    c = long_bitwise((PyLongObject*)a, '^', (PyLongObject*)b);
-    return c;
+    PyLongObject *x = (PyLongObject*)a;
+    PyLongObject *y = (PyLongObject*)b;
+    if (IS_MEDIUM_VALUE(x) && IS_MEDIUM_VALUE(y)) {
+        return _PyLong_FromSDigit(medium_value(x) ^ medium_value(y));
```

Comment:
Some care is needed here. This code isn't correct as written - it's not guaranteed that `x ^ y` fits in a single digit.

Reply:
The same is true for `&`, of course. `|` should be fine as is, but it's probably worth a comment noting that fact.

Reply:
(We should add some regression tests for these cases, too; it's a bit disturbing that the integer tests didn't pick up this bug.)

Reply:
Ah, apologies; I see that `_PyLong_FromSDigit` *does* allow for the case that the input is not a single digit.